src/main/java/com/commercetools/sync/shoppinglists/helpers/ShoppingListBatchValidator.java

```diff
@@ -0,0 +1,222 @@
+                                      @Nonnull final ShoppingListSyncStatistics syncStatistics) {
+        super(syncOptions, syncStatistics);
+    }
+
+    /**
+     * Given the {@link List}&lt;{@link ShoppingListDraft}&gt; of drafts this method attempts to validate
+     * drafts and collect referenced type keys from the draft and return an {@link ImmutablePair}&lt;{@link Set}&lt;
+     * {@link ShoppingListDraft}&gt;,{@link ReferencedKeys}&gt; which contains the {@link Set} of valid drafts and
+     * referenced keys.
+     *
+     * <p>A valid shopping list draft is one which satisfies the following conditions:
+     * <ol>
+     * <li>It is not null</li>
+     * <li>It has a key which is not blank (null/empty)</li>
+     * <li>It has a name which is not null</li>
+     * <li>It has all lineItems AND textLineItems valid</li>
+     * <li>A lineItem is valid if it satisfies the following conditions:
+     * <ol>
+     * <li>It has a SKU which is not blank (null/empty)</li>
+     * </ol>
+     * <li>A textLineItem is valid if it satisfies the following conditions:
+     * <ol>
+     * <li>It has a name which is not blank (null/empty)</li>
+     * </ol>
+     * </li>
+     * </ol>
+     *
+     * @param shoppingListDrafts the shopping list drafts to validate and collect referenced type keys.
+     * @return {@link ImmutablePair}&lt;{@link Set}&lt;{@link ShoppingListDraft}&gt;,
+     * {@link ReferencedKeys}&gt; which contains the {@link Set} of valid drafts and
+     *      referenced keys within a wrapper.
+     */
+    @Override
+    public ImmutablePair<Set<ShoppingListDraft>, ReferencedKeys> validateAndCollectReferencedKeys(
+        @Nonnull final List<ShoppingListDraft> shoppingListDrafts) {
+        final ReferencedKeys referencedKeys = new ReferencedKeys();
+        final Set<ShoppingListDraft> validDrafts = shoppingListDrafts
+            .stream()
+            .filter(this::isValidShoppingListDraft)
+            .peek(shoppingListDraft ->
+                collectReferencedKeys(referencedKeys, shoppingListDraft))
+            .collect(Collectors.toSet());
+
+        return ImmutablePair.of(validDrafts, referencedKeys);
+    }
+
+    private boolean isValidShoppingListDraft(
+        @Nullable final ShoppingListDraft shoppingListDraft) {
+
+        if (shoppingListDraft == null) {
+            handleError(SHOPPING_LIST_DRAFT_IS_NULL);
+        } else if (isBlank(shoppingListDraft.getKey())) {
+            handleError(format(SHOPPING_LIST_DRAFT_KEY_NOT_SET, shoppingListDraft.getName()));
+        } else if (shoppingListDraft.getName() == null || shoppingListDraft.getName().getLocales().isEmpty()) {
+            handleError(format(SHOPPING_LIST_DRAFT_NAME_NOT_SET, shoppingListDraft.getKey()));
+        } else {
+            final List<String> draftErrors = getErrorsInAllLineItemsAndTextLineItems(shoppingListDraft);
+            if (!draftErrors.isEmpty()) {
+                draftErrors.forEach(this::handleError);
```

Comment:
should it be one callback ? or multiple is better because the handleError is calling the `incrementFailed` counter multiple times for a draft?

I would increment the counter one time.

Reply:
@salander85 did you miss this comment ?

Reply:
6d44fbe